src/crafting_gui.cpp

```diff
@@ -258,64 +258,104 @@ const recipe *select_crafting_recipe( int &batch_size )
+
+                                case 'd':
+                                    picking = available_recipes.search( picking, qry_filter_str.substr( 2 ),
+                                                                        recipe_subset::search_type::description_result );
+                                    break;
+
+                                case 'm': {
+                                    auto &learned = g->u.get_learned_recipes();
+                                    std::vector<const recipe *> intersection_result;
+                                    if( query_is_yes( qry_filter_str ) ) {
+                                        std::set_intersection( available_recipes.begin(), available_recipes.end(), learned.begin(),
+                                                               learned.end(), std::back_inserter( intersection_result ) );
+                                    } else {
+                                        std::set_difference( available_recipes.begin(), available_recipes.end(), learned.begin(),
+                                                             learned.end(),
+                                                             std::back_inserter( intersection_result ) );
+                                    }
+                                    // Populate if the first query filter, reduce otherwise
+                                    if( qry_begin == 0 ) {
+                                        picking.swap( intersection_result );
+                                    } else {
+                                        std::vector<const recipe *> col_result;
+                                        std::set_intersection( intersection_result.begin(), intersection_result.end(), picking.begin(),
+                                                               picking.end(), std::back_inserter( col_result ) );
+                                        picking.swap( col_result );
+                                    }
+                                    break;
                                 }
-                                break;
-                            }
 
-                            case 'h': {
-                                std::copy( available_recipes.begin(), available_recipes.end(), std::back_inserter( picking ) );
-                                if( query_is_yes( qry ) ) {
-                                    show_hidden = true;
+                                case 'h': {
+                                    std::vector<const recipe *> available;
+                                    std::copy( available_recipes.begin(), available_recipes.end(), std::back_inserter( available ) );
+                                    std::vector<const recipe *> col_result;
+
+                                    // Populate if the first query filter, reduce otherwise
+                                    if( qry_begin == 0 ) {
+                                        picking.swap( available );
```

Comment:
Using `swap` to assign a `vector` is entirely correct, but no longer necessary in a C++11 world.  It would probably be less confusing to move-assign.

(Similarly for `col_result` below)